crates/ty_python_semantic/src/types.rs

```diff
@@ -4547,6 +4548,36 @@ impl<'db> Type<'db> {
             .to_instance(db)
             .expect("type should be convertible to instance type");
 
+        // Check if __new__ explicitly returns a different type than the instance
+        let new_return_type = new_call_outcome
+            .as_ref()
+            .and_then(|outcome| outcome.as_ref().ok())
+            .and_then(|bindings| {
+                let ret_type = bindings.return_type(db);
+
+                // Skip if return type is unknown/dynamic or None - likely unresolved annotations
+                if ret_type.is_unknown()
+                    || matches!(ret_type, Type::Dynamic(_))
+                    || ret_type.is_none(db)
+                {
+                    return None;
+                }
+
+                // Skip special classes like NamedTuple that have unique constructor behavior
+                // These often return None or other builtin types that should be ignored
+                if matches!(ret_type, Type::KnownInstance(_)) {
+                    return None;
+                }
+
+                // If the return type is different from the instance type, use it
+                // This handles our various cases: int | A, parent class returns, or int
+                if ret_type == instance_ty {
+                    None
+                } else {
+                    Some(ret_type)
+                }
```

Comment:
Seems like we shouldn't really need this equality check? If the types are the same, it won't hurt to use the `__new__` return type.